src/store/products/products.types.ts

```diff
@@ -0,0 +1,45 @@
+// Entities
+
+export type ProductId = number;
+
+export interface Product {
+  product_id: ProductId;
+  name: string;
+  purchase_cost: number;
+}
+
+// Action types
+
+export const PRODUCTS_FETCH_SUCCESS = "@@products/FETCH_SUCCESS";
+export const PRODUCTS_FETCH_BY_ID_SUCCESS = "@@products/FETCH_BY_ID_SUCCESS";
+export const PRODUCTS_ADD_SUCCESS = "@@products/ADD_SUCCESS";
+
+// Products Api
+
+export interface ProductsFetchResponseData {
+  products: Product[];
+}
+export interface ProductsAddParams {
+  name: string;
+}
+export interface ProductsAddResponseData {
+  product: Product;
+}
+
+// Actions
+
+interface ProductsFetchSuccess {
+  type: typeof PRODUCTS_FETCH_SUCCESS;
+  payload: Product[];
+}
+
+interface ProductsAddSuccess {
+  type: typeof PRODUCTS_ADD_SUCCESS;
+  payload: ProductsAddResponseData;
+}
+
+export type ProductsAction = ProductsFetchSuccess | ProductsAddSuccess;
```

Comment:
The syntax of 'X | Y | Z | ...' I haven't seen much of. Intuitively this seems to be saying that there is a sort of higher level type of an action, that could be either a fetch, or an add?

Reply:
union types

https://codingblast.com/typescript-union-types-type-guards-type-aliases/